### PR TITLE
tabular data formatters

### PR DESCRIPTION
**Pull request by colombod**

### CODE REVIEW

File: src/Microsoft.DotNet.Interactive.Formatting/DefaultTabularDataFormatterSet.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+{
+    internal static class DefaultTabularDataFormatterSet
+    {
+        internal static readonly ITypeFormatter[] DefaultFormatters =
+            new ITypeFormatter[]
+            {
+                new TabularDataFormatter<IEnumerable>((context, source, writer) =>
+                {
+
+                    if (source.GetType() != typeof(string))
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Formatting/DefaultTabularDataFormatterSet.cs:16]:
```suggestion
                    if (!(source is string))
```

File: src/Microsoft.DotNet.Interactive.Formatting/TabularDataFieldList.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright(c).NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.DotNet.Interactive.Formatting
+{
+    public class TabularDataFieldList : List<TabularDataSchemaField>
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Formatting/TabularDataFieldList.cs:8]: These types should not be public.

Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Formatting/TabularDataFieldList.cs:8]: i made them public to be used in the data frame extensions, so we can reuse such schema objects

File: src/Microsoft.DotNet.Interactive.Formatting/TabularDataFormatter.cs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+{
+    public static class TabularDataFormatter
+    {
+        // https://specs.frictionlessdata.io/table-schema/#language
+        public const string MimeType = "application/table-schema+json";
+
+        internal static ITypeFormatter[] DefaultFormatters { get; } = DefaultTabularDataFormatterSet.DefaultFormatters;
+
+
+        public static TabularDataSet ToTabularDataSet(this IEnumerable source)
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Formatting/TabularDataFormatter.cs:21]: This method and its return type should probably be internal.

The type of `source` should probably be `IEnumerable<T>` .This doesn't exactly enforce homogeneous contents (which are necessary for valid tabular formatting) but it helps.

File: src/Microsoft.DotNet.Interactive.Formatting/TabularDataSchema.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright(c).NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using Newtonsoft.Json;
+
+namespace Microsoft.DotNet.Interactive.Formatting
+{
+    public class TabularDataSchema
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Formatting/TabularDataSchema.cs:9]: This type should be internal.

File: src/Microsoft.DotNet.Interactive.Formatting/TabularDataSchemaField.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright(c).NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Newtonsoft.Json;
+
+namespace Microsoft.DotNet.Interactive.Formatting
+{
+    public class TabularDataSchemaField
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Formatting/TabularDataSchemaField.cs:8]: This type should be internal.

File: src/Microsoft.DotNet.Interactive.Formatting/TabularDataSet.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
+﻿// Copyright(c).NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.DotNet.Interactive.Formatting
+{
+    public class TabularDataSet
```
Comment (jonsequitur) [inline src/Microsoft.DotNet.Interactive.Formatting/TabularDataSet.cs:9]: This type should be internal.

File: src/Microsoft.DotNet.Interactive.Formatting/TabularDataFieldList.cs (hunk 2 above)
Comment (jonsequitur) [reply src/Microsoft.DotNet.Interactive.Formatting/TabularDataFieldList.cs:8]: I'd have to see the use case but I suspect we should not support this here.

File: src/Microsoft.DotNet.Interactive.Formatting/TabularDataSchemaField.cs (hunk 5 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Formatting/TabularDataSchemaField.cs]: done for the moment

File: src/Microsoft.DotNet.Interactive.Formatting/TabularDataSet.cs (hunk 6 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Formatting/TabularDataSet.cs]: done for the moment

File: src/Microsoft.DotNet.Interactive.Formatting/TabularDataFieldList.cs (hunk 2 above)
Comment (colombod) [reply src/Microsoft.DotNet.Interactive.Formatting/TabularDataFieldList.cs]: making those internal until we need them

**Pull request merged**